Авторизация с помощью CanCanCan

Tekkon opened the pull request.

Commit:
Авторизация с помощью CanCanCan

darialapina commented on app/controllers/application_controller.rb:
```diff
@@ -1,13 +1,21 @@
 
+  rescue_from CanCan::AccessDenied do |exception|
+    redirect_to root_url, alert: exception.message
+  end
```
Сделай обработку ошибки авторизации для всех типов запросов (html, js, json). для js и json можно возвращать 403 статус (forbidden), на клиенте его перехватывать и выводить сообщение

darialapina commented on app/controllers/attachments_controller.rb:
```diff
@@ -3,6 +3,8 @@ class AttachmentsController < ApplicationController
+
   def destroy
     flash[:notice] = 'Only author of question or answer can remove attachment.' unless is_attachmentable_author?
     respond_with @attachment.destroy if is_attachmentable_author?
```
От проверки можно избавиться. И от сообщения выше.
Ну и от метода is_attachmentable_author? тоже.

darialapina commented on app/models/ability.rb:
```diff
@@ -0,0 +1,47 @@
+    can :create, [Question, Answer, Comment, Attachment]
+    can :update, [Question, Answer, Comment], user: user
+    can :edit, [Question, Answer, Comment], user: user
+    can :destroy, [Question, Answer, Comment], user: user
```
Можно записать сразу
`can [:update, :destroy], [Question, Answer, Comment], user: user`
и опустить :edit, так как, по-умолчанию, cancancan создаёт [алиасы](https://github.com/CanCanCommunity/cancancan/wiki/Action-Aliases). Если ты позволяешь пользователю действие update, то будет разрешено и edit (аналогично с create и new)

darialapina commented on spec/models/ability_spec.rb:
```diff
@@ -0,0 +1,124 @@
+      it { should_not be_able_to :update, create(:question, user: another_user), user: user  }
+
+      it { should be_able_to :edit, question, user: user }
+      it { should_not be_able_to :edit, create(:question, user: another_user), user: user  }
```
По сути, получается, что эти проверки лишние (и аналогичные им на :edit ниже).

darialapina commented on spec/models/ability_spec.rb:
```diff
@@ -0,0 +1,124 @@
+      it { should_not be_able_to :destroy, create(:answer, question: question, user: another_user), user: user  }
+
+      it { should be_able_to :mark_favorite, create(:answer, question: question, user: user), user: user  }
+      it { should_not be_able_to :mark_favorite, create(:answer, question: create(:question, user: another_user), user: user), user: user  }
```
в этих проверках вопрос не обязательно должен принадлежать тому же user. Если убрать этот параметр из создания читаемость немного улучшится.

Tekkon replied:
Нельзя убрать параметр user из создания ответа, будет ошибка "Validation failed: User must exist".

darialapina replied:
Лучше в фабрике указать ассоциацию, тогда какой-то пользователь будет создаваться автоматически.
http://www.rubydoc.info/gems/factory_girl/file/GETTING_STARTED.md#Associations

darialapina commented on spec/models/ability_spec.rb:
```diff
@@ -0,0 +1,124 @@
+      it { should be_able_to :create, Attachment }
+
+      it { should be_able_to :destroy, create(:attachment, attachmentable: question, file: file), user: user }
+      it { should be_able_to :destroy, create(:attachment, attachmentable: answer, file: file), user: user }
```
В принципе, тут, для attachment, и ниже, для голосования, достаточно выполнить проверку только для вопроса (ну или только для ответа), как ты для комментариев сделал.

Commit:
Добавил обработку ошибки для разных типов запросов

Commit:
- Упростил запись в ability, убра edit
- Удалил лишние тесты

Commit:
Исправления, связанные с обработкой ошибок авторизации

Commit:
- Добавил ассоциацию на user в фабрике answers
- Убрал указание user для answer в проверках на mark_favorite, т.к. ответ не обязательно должен принадлежать тому же пользователю, который вызывает экшен mark_favorite
